Source/Metal/DeviceMTL.mm

```diff
@@ -0,0 +1,490 @@
+#include "BufferMTL.h"
+#include "CommandQueueMTL.h"
+#include "TextureMTL.h"
+#include "FenceMTL.h"
+#include "MemoryMTL.h"
+#include "PipelineLayoutMTL.h"
+#include "PipelineMTL.h"
+
+using namespace nri;
+
+static bool FindMTLGpuFamily(id<MTLDevice> device,
+                             const MTLGPUFamily *families, size_t len,
+                             MTLGPUFamily* current) {
+    for(size_t i = 0; i < len; i++) {
+        if( [device supportsFamily: families[i]]) {
+            *current = families[i];
+            return true;
+        }
+    }
+    return false;
+}
+
+// referenced from Molten VK
+static uint32_t GetEntryProperty(io_registry_entry_t entry, CFStringRef propertyName) {
+
+    uint32_t value = 0;
+
+    CFTypeRef cfProp = IORegistryEntrySearchCFProperty(entry,
+                                                       kIOServicePlane,
+                                                       propertyName,
+                                                       kCFAllocatorDefault,
+                                                       kIORegistryIterateRecursively |
+                                                       kIORegistryIterateParents);
+    if (cfProp) {
+        const uint32_t* pValue = reinterpret_cast<const uint32_t*>(CFDataGetBytePtr((CFDataRef)cfProp));
+        if (pValue) { value = *pValue; }
+        CFRelease(cfProp);
+    }
+
+    return value;
+}
+
+DeviceMTL::DeviceMTL(const CallbackInterface& callbacks, const StdAllocator<uint8_t>& stdAllocator)
+: DeviceBase(callbacks, stdAllocator) {
+    m_Desc.graphicsAPI = GraphicsAPI::MTL;
+    m_Desc.nriVersionMajor = NRI_VERSION_MAJOR;
+    m_Desc.nriVersionMinor = NRI_VERSION_MINOR;
+}
+
+
+void DeviceMTL::Destruct() {
+    for (uint32_t i = 0; i < m_CommandQueues.size(); i++)
+        Destroy(GetStdAllocator(), m_CommandQueues[i]);
+    Destroy(GetStdAllocator(), this);
+}
+
+DeviceMTL::~DeviceMTL() {
+
+}
+
+
+//FormatSupportBits DeviceMTL::GetFormatSupport(const Device& device, Format format) {
+//  int currentFamily = HIGHEST_GPU_FAMILY;
+//  for (; currentFamily >= (int)MTLGPUFamilyApple1; currentFamily--) {
+//    if ([m_Device supportsFamily:(MTLGPUFamily)currentFamily]) {
+//      highestAppleFamily = (MTLGPUFamily)currentFamily;
+//      break;
+//    }
+//  }
+//}
+
+
+static id<MTLFunction> initShaderFromSource(id<MTLDevice> dev, NSString* src, NSString* method) {
+    id<MTLFunction> mtlFunc = nil;
+    NSError* err = nil;
+    
+    MTLCompileOptions* options = [MTLCompileOptions alloc];
+    id<MTLLibrary> mtlLib = [dev newLibraryWithSource: src
+                                                   options: options
+                                                     error: &err];    // temp retain
+    NSCAssert(err,  @"Failed to load Metal shader library %@", err );
+    
+    return [mtlLib newFunctionWithName: method];
+}
+
+
+id<MTLRenderPipelineState> DeviceMTL::GetClearPipeline(ClearDesc* desc, size_t numFormats) {
```

Comment:
for this chunk of logic I need a way to cache pipelines for clearing. Any clue how this might be implemented for this portion of the code.

Reply:
(sorry, hard to comment on other questions due to lack of expertise)

Reply:
i see, ill have to do my due diligence. im somewhat close to getting something working. might take a bit longer before i finalize anything.